turbopack/crates/turbo-tasks/src/parallel.rs

```diff
@@ -0,0 +1,315 @@
+//! Parallel for each resp. map running in the current tokio thread pool maintaining turbo tasks and
+//! tracing context.
+//!
+//! This avoid the problem of sleeping threads with mimalloc when using rayon in combination with
+//! tokio. It also avoid having multiple thread pools.
+
+use std::{sync::LazyLock, thread::available_parallelism};
+
+use crate::{scope::scope_and_block, util::into_chunks};
+
+/// Calculates a good chunk size for parallel processing based on the number of available threads.
+/// This is used to ensure that the workload is evenly distributed across the threads.
+fn good_chunk_size(len: usize) -> usize {
+    static GOOD_CHUNK_COUNT: LazyLock<usize> =
+        LazyLock::new(|| available_parallelism().map_or(16, |c| c.get() * 4));
+    let min_chunk_count = *GOOD_CHUNK_COUNT;
+    len.div_ceil(min_chunk_count)
+}
+
+pub fn for_each<'l, T, F>(items: &'l [T], f: F)
+where
+    T: Sync,
+    F: Fn(&'l T) + Send + Sync,
+{
+    let len = items.len();
+    if len == 0 {
+        return;
+    }
+    if len == 1 {
+        for item in items {
+            f(item);
+        }
+        return;
+    }
```

Comment:
I think this should be roughly equivalent.

```suggestion
    let len = items.len();
    if len <= 1 {
        for item in items {
            f(item);
        }
        return;
    }
```